### PR TITLE
Add MIP driver for Teensy 4.1 board

### PR DESCRIPTION
**Pull request by psykokwak-com**
Working implementation of MIP driver for Teensy 4.1 board

### CODE REVIEW

**Comment by psykokwak-com** [comment]
Can you explain me how to add a MIP driver into the mongoose build system ?

**Comment by psykokwak-com** [comment]
Curl performance test is not so bad (with chunked payload of 1390Byte) 👍 
```
C:\Users\Jyce\Desktop>curl http://192.168.123.10/ -o test.bin
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  269M    0  269M    0     0  11.2M      0 --:--:--  0:00:24 --:--:-- 11.2M
```

**Comment by cpq** [comment]
@psykokwak-com thank you!

Could you please:
- Squash all commits into one. We allow only one commit per PR
- Confirm you're contributing your code under our license (copyright Cesanta, dual GPLv2/commercial)
- Sign the [CLA](https://cesanta.com/cla.html) linked from the project's README

**Comment by psykokwak-com** [comment]
Ok. I squashed the commits about this subject (I will do another PR for the others one).
I confirm I'm contributing the code under your license (copyright Cesanta, dual GPLv2/commercial)
CLA signed.

PS : I let you integrate this MIP driver into the correct mongoose build architecture.

**Comment by ssilverman** [comment]
I believe you also need to include the copyright and MIT license at the top, per the MIT license terms, because it looks like it uses a “substantial portion” of the code. From the license:

> The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

Plus whatever Paul Stoffregen requires, because the _QNEthernet_ low-level stuff was based on his code.

**Comment by psykokwak-com** [comment]
it's starting to get complicated.
I leave it up to you to make the changes you need to make to allow the driver to be added to the project.
For my part, I have no particular license condition.

**Comment by psykokwak-com** [comment]
I understand. No problem.
For information, I fixed some issues from the QNEthernet low level code. I let you check.
Then I use this board : https://www.skpang.co.uk/products/teensy-4-1-triple-can-board-with-240x240-lcd-and-ethernet

File: mip/driver_teensy41.c [hunk 1]
```diff
@@ -0,0 +1,579 @@
+#include "driver_teensy41.h"
+
+// Author: Jean-Charles Mammana
```
Comment (ssilverman) [inline mip/driver_teensy41.c:3]: I want it on record that Jean-Charles Mammana is _not_ the author of this code. He may have made some modifications here, but this is not his code.

Comment (iceberg1369) [reply mip/driver_teensy41.c:3]: I like to see mongoose on teensy, good work

**Comment by cpq** [comment]
@psykokwak-com @ssilverman Gents, your work is appreciated.
I'd love to integrate this, however I want to be crystal clear about the license terms.
Unless it is BSD/MIT/public domain all through, we cannot merge it, cause Mongoose is GPLv2/commercial.
If there are other parties to get consent from, it is too risky and complicated, and it is a show stopper for us.

**Comment by iceberg1369** [comment]
> 

is is w5500

**Comment by ssilverman** [comment]
> @psykokwak-com @ssilverman Gents, your work is appreciated. I'd love to integrate this, however I want to be crystal clear about the license terms. Unless it is BSD/MIT/public domain all through, we cannot merge it, cause Mongoose is GPLv2/commercial. If there are other parties to get consent from, it is too risky and complicated, and it is a show stopper for us. Please clarify, who owns the copyright for this change, and under which terms.
> 
> Quick question: does teensy 4.1 use the same Ethernet controller as [rt1020](https://mongoose.ws/documentation/#mimxrt1020-evk) ? We almost have the driver for rt1020 ready.

I'm not attached to Mongoose support on the Teensy 4.1. I didn't make this PR. My issue is that @psykokwak-com put his name as the author on my code (i.e. `// Author: Jean-Charles Mammana`). This is not acceptable, especially since only very minor (and incorrect) changes were made. In addition, he didn't follow my licensing terms. The licence states:

> The above copyright notice and this permission notice shall be included in all
> copies or substantial portions of the Software.

He modified the file to say that it's "based on" my code, and that's not correct. It's a straight-up copy of my code. He didn't "base" it on anything; that's misleading. (Especially since none of the changes are necessary or correct.)

In short: Don't copy my code, say it's yours, and then attempt to give others permission to use it without following the licence. I _actually_ based my code on others' code, _and_ reached out to them to find out what terms they'd like. This is not @psykokwak-com's code, this is mine, and I hold the copyright. He does not, not does he have permission to redistribute it outside the licensing terms.

**Comment by iceberg1369** [comment]
> 

could you make a dedicated pull request for mongoose. it is really sad to miss mongoose on teensy 4.1

**Comment by ssilverman** [comment]
> could you make a dedicated pull request for mongoose. it is really sad to miss mongoose on teensy 4.1

I’m personally not interested in having my code being part of this project unless I’m compensated for it. Mongoose is commercial.

**Comment by scaprile** [comment]
Gentlemen, 

> We will do an independent implementation from scratch to avoid any licensing issues, as we do for the rest of our drivers. We'll do a bare metal , no-sdk example as part of it, as we do for the other bare metal examples for other architectures.

Mongoose is dual-licensed:
https://github.com/cesanta/mongoose/blob/d5b5cec7970473d0526262ac42c93ea8a98f0699/LICENSE#L1-L16

**Comment by cpq** [comment]
Since there is copyright issue, we can't merge this.
Though I think we'll have driver for the rt1020 done soon. We'll test it on MIMXRT1020-EVK, and chances are it'll work on Teensy 4.1 - we'll test it there, too.

@ssilverman as @scaprile  has already pointed out, this project is dual licensed. It's free for open source projects.
Said that, if you're willing to contribute to this project and be compensated, please contact us at support@cesanta.com - I'd be happy to discuss.
